File: Java/src/LargestValuesFromLabels.java
import java.util.Arrays;
import java.util.HashMap;


public class LargestValuesFromLabels {

    class Node {
        int v;
        int l;
        Node(int v, int l){
            this.v = v;
            this.l = l;
        }
    }

    public int largestValsFromLabels(int[] values, int[] labels, int num_wanted, int use_limit) {
        
        Node[] l = new Node[values.length];
        for(int i = 0; i < values.length; i++){
            l[i] = new Node(values[i], labels[i]);
        }

        Arrays.sort(l, (a, b)->{
            return b.v - a.v;
        });

        HashMap<Integer, Integer> map = new HashMap<>();

        int res = 0;
        int count = 0;
        for(int i = 0; i < l.length; i++){
            if(map.containsKey(l[i].l) && map.get(l[i].l) == use_limit) continue;
            map.put(l[i].l, map.getOrDefault(l[i].l, 0)+1);
            count++;
            res += l[i].v;
            if(count == num_wanted) break;
        }
    
        return res;
    }


    public static void main(String[] args) {
        LargestValuesFromLabels test = new LargestValuesFromLabels();
        int[] values = {9,8,8,7,6};
        int[] labels = {0,0,0,1,1};
        System.out.println(test.largestValsFromLabels(values, labels, 3, 1));
    }

}

File: Python/LargestValuesFromLabels.py
from typing import List

class Solution:
    def largestValsFromLabels(self, values: List[int], labels: List[int], num_wanted: int, use_limit: int) -> int:
        
        l = []
        for i in range(len(values)):
            l.append((values[i], labels[i]))
        l = sorted(l, key=lambda a: a[0])

        count = {}
        n = 0
        res = 0
        print(l)
        for i in range(len(l)-1, -1, -1):
            if l[i][1] in count and count[l[i][1]] == use_limit:
                continue
            
            if l[i][1] not in count:
                count[l[i][1]] = 1
            else:
                count[l[i][1]] += 1

            n += 1
            res += l[i][0]

            print(count)
            if n == num_wanted:
                break

        return res


test = Solution()
print(test.largestValsFromLabels([5,4,3,2,1], [1,1,2,2,3], 3, 1))






